### DISCUSSION TITLE
Empty table sub-captions when rendering to PDF (kableExtra)

### DISCUSSION THREAD

**Discussion by nclJoshCowley** [quarto-dev/quarto-cli#4206] [Q&A]
Hi all,

I'm attempting to use the `kableExtra` package to make two sub-tables but empty sub-captions are being created.

An easy workaround is to split the tables into Table 1 and Table 2 (likely makes more sense anyway)
but thought I'd share this behaviour I don't understand.

In my project I use other `kableExtra` functions but consider this toy example to show the issue.

````
---
title: "kbltest"
format: pdf
---

```{r}
#| label: tb-knitr
#| tbl-cap: "Main caption"
#| tbl-subcap:
#|   - "y"
#|   - "z"
#| layout-ncol: 2

library(knitr)
knitr::kable(data.frame(y = rnorm(n = 5)))
knitr::kable(data.frame(z = rnorm(n = 5)))
```

\newpage

```{r}
#| label: tb-kableExtra
#| tbl-cap: "Main caption"
#| tbl-subcap:
#|   - "y"
#|   - "z"
#| layout-ncol: 2

library(kableExtra)
kableExtra::kable(data.frame(y = rnorm(n = 5)))
kableExtra::kable(data.frame(z = rnorm(n = 5)))
```
````

In the first chunk, knitr will create two tables, 1.(a) and 1.(b).

For the second chunk, the kable function is seemingly identical and a re-export of the knitr package.
However, it leads to duplication of captions where 1.(a) and 1.(b) are blank whereas 1.(c) and 1.(d) are the desired subcaps.

Off-topic, what are people's goto for rendering tables to PDF in Quarto?
I like `gt` but not sure how good the LaTeX functionality is.

**Comment by cwickham** [comment]
A look at the intermediate markdown suggests kableExtra is doing something to how the format is detected. The knitr tables end up in pipe format, and the kableExtra ones in latex.

<details>
<summary>Intermediate markdown</summary>

````
---
title: "kbltest"
format: pdf
keep-md: true
---


::: {.cell layout-ncol="2" tbl-cap='Main caption' tbl-subcap='["y","z"]'}

```{.r .cell-code}
library(knitr)
knitr::kable(data.frame(y = rnorm(n = 5)))
```

::: {.cell-output-display}
|          y|
|----------:|
|  1.4823228|
| -0.2169394|
|  0.2413181|
| -1.3778683|
|  0.8386497|
:::

```{.r .cell-code}
knitr::kable(data.frame(z = rnorm(n = 5)))
```

::: {.cell-output-display}
|          z|
|----------:|
| -0.4361908|
| -0.7627225|
|  0.0763406|
| -0.6971404|
|  2.5104861|
:::
:::


\newpage


::: {.cell layout-ncol="2" tbl-cap='Main caption' tbl-subcap='["y","z"]'}

```{.r .cell-code}
library(kableExtra)
kableExtra::kable(data.frame(y = rnorm(n = 5)))
```

::: {.cell-output-display}
\begin{tabular}{r}
\hline
y\\
\hline
1.110585\\
\hline
1.753608\\
\hline
0.771072\\
\hline
1.620340\\
\hline
3.453606\\
\hline
\end{tabular}
:::

```{.r .cell-code}
kableExtra::kable(data.frame(z = rnorm(n = 5)))
```

::: {.cell-output-display}
\begin{tabular}{r}
\hline
z\\
\hline
1.0128286\\
\hline
-0.9067493\\
\hline
-0.4840154\\
\hline
-1.8765782\\
\hline
-0.0510191\\
\hline
\end{tabular}
:::
:::

````

</details>

One fix is to explicitly specify the format argument in `kableExtra::kable()`:

````
---
title: "kbltest"
format: pdf
keep-md: true
---


```{r}
#| label: tb-knitr
#| tbl-cap: "Main caption"
#| tbl-subcap:
#|   - "y"
#|   - "z"
#| layout-ncol: 2

knitr::kable(data.frame(y = rnorm(n = 5)))
knitr::kable(data.frame(z = rnorm(n = 5)))
```

\newpage

```{r}
#| label: tb-kableExtra
#| tbl-cap: "Main caption"
#| tbl-subcap:
#|   - "y"
#|   - "z"
#| layout-ncol: 2

library(kableExtra)
kableExtra::kable(data.frame(y = rnorm(n = 5)), format = "pipe")
kableExtra::kable(data.frame(z = rnorm(n = 5)), format = "pipe")
```

````

I'm not sure exactly why this happens, but it is worth noting if you flip the order of your tables so the kableExtra ones are first, the kable ones end up with the same problem.

**Reply by cderv** [reply]
**kableExtra** will change the default format used by `knitr::kable()` when the package is loaded. This is what explains indeed the differences. This is here in there code 
https://github.com/haozhu233/kableExtra/blob/master/R/zzz.R#L21-L22
https://github.com/haozhu233/kableExtra/blob/292f60715959ea952ff25a8aedfd782793b15f7b/R/auto_set_format.R#L1-L7

They set to `latex` when rendering to PDF or otherwise to HTML table. **kableExtra** does not help customize Markdown table if I understand correctly. So no difference with using `knitr::kable()` default feature. 

Setting `options(kableExtra.auto_format=FALSE)` can deactivate this behavior thought, but I don't know what is lost then regarding **kableExtra** features.

It is explained in the package documentation https://haozhu233.github.io/kableExtra/ for PDF and HTML - you'll found more details there.


The question here regarding `tbl-subcap` feature is: Does it support multiple raw LaTeX table ? It seems not for now.

**Reply by nclJoshCowley** [reply]
Thanks both for the detailed explanation, I was incorrectly assuming kable went straight to LaTeX and forgot about markdown / pipe tables.

> The question here regarding `tbl-subcap` feature is: Does it support multiple raw LaTeX table ? It seems not for now.

This would explain why I was seeing similar behaviour with `gt::as_latex(...)`

**Comment by cscheid** [comment]
(Just in passing, I'll note that if you want crossrefs to work with these figures, you'll need your labels to start with `tbl-` rather than `tb-`.)
